File: packages/flutter-dart/lib/appkit_smart_account/user_operation.dart
/// ERC-4337 User Operation
class UserOperation {
  final String sender;
  final int nonce;
  final List<int> initCode;
  final List<int> callData;
  final int callGasLimit;
  final int verificationGasLimit;
  final int preVerificationGas;
  final int maxFeePerGas;
  final int maxPriorityFeePerGas;
  final List<int> paymasterAndData;
  final List<int> signature;

  const UserOperation({
    required this.sender,
    required this.nonce,
    required this.initCode,
    required this.callData,
    required this.callGasLimit,
    required this.verificationGasLimit,
    required this.preVerificationGas,
    required this.maxFeePerGas,
    required this.maxPriorityFeePerGas,
    required this.paymasterAndData,
    required this.signature,
  });

  Map<String, String> toHexParams() => {
        'sender': sender,
        'nonce': '0x${nonce.toRadixString(16)}',
        'initCode': _bytesToHex(initCode),
        'callData': _bytesToHex(callData),
        'callGasLimit': '0x${callGasLimit.toRadixString(16)}',
        'verificationGasLimit': '0x${verificationGasLimit.toRadixString(16)}',
        'preVerificationGas': '0x${preVerificationGas.toRadixString(16)}',
        'maxFeePerGas': '0x${maxFeePerGas.toRadixString(16)}',
        'maxPriorityFeePerGas': '0x${maxPriorityFeePerGas.toRadixString(16)}',
        'paymasterAndData': _bytesToHex(paymasterAndData),
        'signature': _bytesToHex(signature),
      };

  static String _bytesToHex(List<int> bytes) {
    if (bytes.isEmpty) return '0x';
    return '0x${bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join()}';
  }
}

File: packages/ios-swift/Sources/CinacoinAppKitSmartAccount/UserOperation.swift
import Foundation

/// ERC-4337 User Operation
public struct UserOperation: Codable, Sendable {
    public let sender: String
    public let nonce: UInt64
    public let initCode: Data
    public let callData: Data
    public let callGasLimit: UInt64
    public let verificationGasLimit: UInt64
    public let preVerificationGas: UInt64
    public let maxFeePerGas: UInt64
    public let maxPriorityFeePerGas: UInt64
    public let paymasterAndData: Data
    public let signature: Data
    
    public init(
        sender: String,
        nonce: UInt64,
        initCode: Data,
        callData: Data,
        callGasLimit: UInt64,
        verificationGasLimit: UInt64,
        preVerificationGas: UInt64,
        maxFeePerGas: UInt64,
        maxPriorityFeePerGas: UInt64,
        paymasterAndData: Data,
        signature: Data
    ) {
        self.sender = sender
        self.nonce = nonce
        self.initCode = initCode
        self.callData = callData
        self.callGasLimit = callGasLimit
        self.verificationGasLimit = verificationGasLimit
        self.preVerificationGas = preVerificationGas
        self.maxFeePerGas = maxFeePerGas
        self.maxPriorityFeePerGas = maxPriorityFeePerGas
        self.paymasterAndData = paymasterAndData
        self.signature = signature
    }
    
    /// Encode to hex string for RPC
    public func toHexParams() -> [String: String] {
        return [
            "sender": sender,
            "nonce": "0x\(String(nonce, radix: 16))",
            "initCode": initCode.hexString,
            "callData": callData.hexString,
            "callGasLimit": "0x\(String(callGasLimit, radix: 16))",
            "verificationGasLimit": "0x\(String(verificationGasLimit, radix: 16))",
            "preVerificationGas": "0x\(String(preVerificationGas, radix: 16))",
            "maxFeePerGas": "0x\(String(maxFeePerGas, radix: 16))",
            "maxPriorityFeePerGas": "0x\(String(maxPriorityFeePerGas, radix: 16))",
            "paymasterAndData": paymasterAndData.hexString,
            "signature": signature.hexString
        ]
    }
}

extension Data {
    var hexString: String {
        "0x" + map { String(format: "%02x", $0) }.joined()
    }
}
